pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+        if self.use_flat and flatimages is None:
+            msgs.error('Flat-field corrections requested but no flat-field images generated '
+                       'or provided.  Make sure you have flat-field images in your PypeIt file!')
+
+        # Apply additive corrections.  The order matters and is fixed.
+        #   - Subtract any fixed pattern defined for the instrument.  NOTE: This
+        #     step *must* be done before use_overscan
+        if self.par['use_pattern']:
             self.subtract_pattern()
-        if par['use_overscan']:
+
+        #   - Estimate the readnoise using the overscan regions.  NOTE: I'm
+        #     assuming this needs to be done *after* the sine pattern has been
+        #     subtracted.
+        if not np.all(self.ronoise > 0):
+            self.estimate_readnoise()
+
+        #   - Get detector variance image.  Starts with the shape and
+        #     orientation of the raw image, and the units are in ADU to match
+        #     the current units of the image being processed.  Note that the
+        #     readnoise variance image is kept separate because it is used again
+        #     when adding the shot noise.
+        self.rn2img = self.build_rn2img(units='ADU')
+        self.var = self.rn2img.copy()
+
+        #   - Subtract the overscan.  Uncertainty from the overscan subtraction
+        #     is added to the variance.
+        if self.par['use_overscan']:
             self.subtract_overscan()
-        if par['trim']:
+
+        # TODO: Do we need to keep trim and orient as optional?
+        #   - Trim to the data region.  This trims the image, rn2img, var, and
+        #     datasec_img.
+        if self.par['trim']:
             self.trim()
-        if par['orient']:
+
+        #   - Re-orient to PypeIt convention. This re-orients the image, rn2img,
+        #     var, and datasec_img.
+        if self.par['orient']:
             self.orient()
-        if par['use_biasimage']:  # Bias frame, if it exists, is *not* trimmed nor oriented
+            
+        #   - Check the shape of the bpm
+        if self.bpm.shape != self.image.shape:
+            msgs.error(f'Bad-pixel mask has incorrect shape: found {self.bpm.shape}, expected '
+                       f'{self.image.shape}.  The shape must match a trimmed and oriented '
+                       'PypeIt-processed image.')
+            
+        #   - Subtract master bias
+        if self.par['use_biasimage']:
+            # Bias frame.  Shape and orientation must match *processed* image,
+            # and the units *must* be ADU.  Uncertainty from the bias
+            # subtraction is added to the variance.
             self.subtract_bias(bias)
-        if par['use_darkimage']:  # Dark frame, if it exists, is TODO:: check: trimmed, oriented (and oscan/bias subtracted?)
-            self.subtract_dark(dark)
-        if par['apply_gain']:
+            
+        #   - Convert from ADU to electrons/counts.  This converts the image,
```

Comment:
I disagree with the order of operations here. You should be converting everything into electrons at the beginning, and then computing the noise models in electrons. I don't see any compelling argument to do this any other way. The noise models are defined in terms of photon counting statistics, which are only valid when images are in electrons. Also read noise is defined in terms of electrons, so it is counter-intuitive to every compute a rn2 image in ADU.

Reply:
Yep, conversion from ADU to electron counts is now done first.